| <img src="https://raw.githubusercontent.com/abcdbug/dbug/master/WCM_logo.png" alt="WCM" style="height: 20px;"/> | [Applied Bioinformatics Core](https://abc.med.cornell.edu) |
|---------------|---------------|

# RNA-seq course July 2019

Here, you can find (some) of the code that we used during class.
The first two days are mostly bash commands covering raw reads and alignment.
Day 3 and 4 are based on `Rmarkdown` files.
None of these file may be completely fool proof since I have not tested them extensively, but they should give you a good idea of what we did during class.

The full 80+pages course notes can be found on our [server](http://chagall.med.cornell.edu/RNASEQcourse/Intro2RNAseq.pdf) or on [zenodo](https://zenodo.org/record/4390117).

Before there was [MultiQC](http://multiqc.info/), you had to do your QC output wrangling yourself. [This repo for previous class versions](https://github.com/friedue/course_RNA-seq2015) contains additional scripts that show you how some of the images (e.g., the bar charts of read alignment percentages) within the report were produced.

-----------------------------------------
[return to course material website at WCM](http://chagall.med.cornell.edu/RNASEQcourse/)
